File: 208-implement-trie-prefix-tree/implement-trie-prefix-tree.cpp
struct TrieNode{
    unordered_map<char,TrieNode*> children;
    bool endofword=false;
};
class Trie {
public:
    TrieNode* root=new TrieNode();
    Trie() {
        
    }
    
    void insert(string word) {
        TrieNode* node=root;
        for(auto c:word){
            if(node->children.find(c)==node->children.end()){
                node->children[c]=new TrieNode();
            }
            node=node->children[c];
        }
        node->endofword=true;
    }
    
    bool search(string word) {
        TrieNode* node=root;
        for(auto c:word){
            if(node->children.find(c)==node->children.end()) return false;
            node=node->children[c];
        }
        return node->endofword;
    
    }
    
    bool startsWith(string prefix) {
        TrieNode* node=root;
        for(auto c:prefix){
            if(node->children.find(c)==node->children.end()) return false;
            node=node->children[c];
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: 208-implement-trie-prefix-tree/implement-trie-prefix-tree.java
class Trie {

    class Node {
        Map<Character,Node> map;
        boolean endofflag;
        public Node(){
            map=new HashMap<>();
            endofflag=false;
        }
    }

    Node root;

    public Trie() {
        root=new Node();
    }
    
    public void insert(String word) {
        Node curr=root;
        for(int i=0;i<word.length();i++){
            char c=word.charAt(i);
            if(!curr.map.containsKey(c)){
                curr.map.put(c,new Node());
            }
            curr=curr.map.get(c);
        }
        curr.endofflag=true;
    }
    
    public boolean search(String word) {
        Node curr=root;
        for(int i=0;i<word.length();i++){
            char ch=word.charAt(i);
            if(!curr.map.containsKey(ch)){
                return false;
            }
            curr=curr.map.get(ch);
        }
        return curr.endofflag;
    }
    
    public boolean startsWith(String prefix) {
        Node curr=root;
        for(int i=0;i<prefix.length();i++){
            char ch=prefix.charAt(i);
            if(!curr.map.containsKey(ch)){
                return false;
            }
            curr=curr.map.get(ch);
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */